### ISSUE TITLE
libnghttp2 degenerative behavior possible when HTTP/2 window size exhausted

### ISSUE THREAD

**Issue by gstrauss** [nghttp2/nghttp2#1722]
libnghttp2 degenerative behavior possible when HTTP/2 window size exhausted

curl using libnghttp2 to upload a file over HTTP/2 may in some scenarios end up sending h2 DATA frames with 1-byte each for most of the DATA frames, leading to very high CPU utilization and extremely slow uploads:

A fast client machine will quickly exhaust the h2 send window (default SETTINGS_INITIAL_WINDOW_SIZE 65535 (64k-1)) by sending large DATA frames (default SETTINGS_MAX_FRAME_SIZE 16384 (16k)) in a sequence of size: (16384, 16384, 16384, 16383).

As a server receives h2 DATA frames, a server will typically send h2 WINDOW_UPDATE with exactly the number of bytes received in h2 DATA frame in order to replenish the client send window size.

A fast client machine will quickly send the next block of data immediately upon receiving h2 WINDOW_UPDATE frame from server, once again exhausing the h2 send window.

Now, if the client uploading a large file always fills the window size, then h2 WINDOW_UPDATE frames from the server will also typically be large.

Unfortunately, if the fast client chooses not to fill the window size, then a smaller DATA frame is injected into the sequence, e.g. if the client uses a 64k buffer, the fast client may send DATA frames in a sequence of sizes: (16384, 16384, 16384, 16383, 1) before reusing the 64k buffer to read the next 64k block to be sent.

This can -- and, in practice, does (!) -- lead to severe degenerative behavior, e.g. one person reported a 200 MB upload which took 1-2 mins with HTTP/1.1 now took about 30 mins with HTTP/2.  https://redmine.lighttpd.net/issues/3089

The fast client continues injecting an additional 1-byte DATA frame every 64k eventually leading to almost every frame being a 1-byte DATA frame!
```
  --> (16384, 16384, 16384, 16383)
  --> (1, 16383, 16384, 16384, 16383, 1)
  --> (1, 1, 16382, 16384, 16384, 16383, 1)
  --> (1, 1, 1, 16381, 16384, 16384, 16383, 1)
  --> (1, 1, 1, 1, 16380, 16384, 16384, 16383, 1)
  --> (1, 1, 1, 1, 1, 16379, 16384, 16384, 16383, 1)
  ...
```

```
Consider the initial send from the fast client to fill the 65535 window size:
  --> (16384, 16384, 16384, 16383)
Now, when the client receives the first WINDOW_UPDATE from the server
              <-- (16384)
the fast client sends 1 byte to finish the client 64k block, and then
continues with the next 64k block.  Since the window size was 16384:
  --> (1, 16383)
The server continues sending WINDOW_UPDATE for previously sent frames
              <-- (16384)
and the client immediately sends more data
  --> (16384)
continuing
              <-- (16384)
  --> (16384)
              <-- (16383)
  --> (16383)
              <-- (1)
  --> (1)
The client still has 1 byte left to send in its 64k block and so injects
another 1-byte DATA frame upon receiving the next WINDOW_UPDATE from server
before the client starts its next 64k block, this time sending only 16382.
              <-- (16383)
  --> (1, 16382)
```

I believe there is a design flaw in the defaults chosen in RFC 7540 that SETTINGS_INITIAL_WINDOW_SIZE 65535 (64k-1) is not a multiple of SETTINGS_MAX_FRAME_SIZE 16384 (16k).  It should have been a clean multiple.

I think that client and servers would do well to unconditionally send SETTINGS_INITIAL_WINDOW_SIZE 65536 in the HTTP/2 connection preface, and to pre-emptively increase session window size by sending WINDOW_UPDATE with +1 for id 0.

**However, for the vast deployments out there which use the default window sizes, or use libnghttp2 to manage the windows, is there anything that can be changed in libnghttp2 to avoid the degenerative behavior described above?**

Potentially related (very old) issues: #714 #730

**Comment by gstrauss** [comment]
@bagder: this degenerative behavior scenario might explain some high CPU use and slow uploads using curl and HTTP/2

**Comment by tatsuhiro-t** [comment]
I'm not sure this is libnghttp2 issue.  When 16384 window available, client can send 16384 DATA frame instead of 1 + 16383 by coalescing data blocks.

**Comment by gstrauss** [comment]
I, too, am not sure this is directly a libnghttp2 issue, but libnghttp2 is a library and what has been reported above is a real scenario (not theoretical) which occurs with a simple application reusing a buffer, and feeding libnghttp2 which manages the h2 window size.

**Comment by tatsuhiro-t** [comment]
If you really want to avoid 1 byte send, try coalescing data block.  Then if you also avoid to 1 byte write because of window size, you can use nghttp2_session_get_stream_remote_window_size and nghttp2_session_get_remote_window_size and wait for window size to grow, but you never know how much window size server gives you.  The initial window size does not help, because it is just an initial value.  Server can increase or decrease the window size as it wants.  So if you want the larger window size than server is willing to allow, then you will see the dead lock.  So I think it would be better send data by fully using window size available without waiting more size.

**Comment by gstrauss** [comment]
> So I think it would be better send data by fully using window size available without waiting more size.

I agree, specifically with "fully using window size".  However, the reason I reported this to nghttp2 is because it is not unreasonable for a simple application using libnghttp2 to reuse a 64k buffer, and to try to send the full 64k buffer in 5 DATA frames of sizes (16384, 16384, 16384, 16383, 1) before going to reuse that 64k buffer to read more input from, say, a pipe.  In doing so, the application might inadvertently fall into the degenerative behavior reported above.

Is there something that libnghttp2 might do to detect and prevent this behavior?  If libnghttp2 can know that the client has xxx (> some low watermark) data ready to send, then might libnghttp2 defer sending the next DATA frame until there is a minimum of (low watermark) available in window size?  Could this be a configuration setting to enable or disable?  As you pointed out, there is a chance of deadlock/timeout if server never increases window size, but if client has more than that data thta the client wants to send, then that mismatch is its own problem.  In other words, I think that some applications would enable a window size low watermark if such a feature were available in libnghttp2.  (I recognize that there would also be the chance of stream starvation if other streams continually used up the available window, not allowing it to reach the low watermark.)

I have not yet tracked down why the behavior I am seeing occurs *sometimes* with curl.

**Comment by gstrauss** [comment]
In the following, the same HTTP request uploading a 100MB file is sent via HTTP/2 with and without TLS, and with HTTP/1.1 with TLS for comparison.

Server: lighttpd 1.4.64, which uses HTTP/2 default window sizes (65535)
        Requests are handled by a trivial CGI script which collects the request body
prep: `dd if=/dev/urandom bs=1M count=100 >rand100m`

Client: curl (which uses libnghttp2)
```
time curl -s -k --data-binary '@-' https://localhost/ <rand100m
real    0m38.472s
user    0m14.981s
sys     0m18.451s
```
```
time curl --http1.1 -s -k --data-binary '@-' https://localhost/ <rand100m
real    0m0.331s
user    0m0.126s
sys     0m0.103s
```
```
time curl --http2-prior-knowledge -s --data-binary '@-' http://localhost/ <rand100m
real    0m0.334s
user    0m0.064s
sys     0m0.170s
```

Client: nghttp
```
time nghttp -d - https://localhost/ <rand100m
real    0m0.363s
user    0m0.158s
sys     0m0.129s
```
```
time nghttp -d - http://localhost/ <rand100m
real    0m0.235s
user    0m0.030s
sys     0m0.120s
```

While the above all occurs on localhost, for some reason one of the scenarios has timing / kernel task scheduling / other differences which fall into the degenerative scenario, taking 38 seconds instead of < 1/2 second.

I can provide the debug output from libnghttp2 if you would like to see that libnghttp2 is sending many, many, many 1-byte DATA frames.

**Comment by tatsuhiro-t** [comment]
I have no plan to provide configuration for watermark and reusing fixed sized buffers as the way you described.  Application that cares about the performance should coalesce buffers.  That is how libnghttp2 is designed unfortunately.  For watermark, I think application can implement it using nghttp2_session_get_stream_remote_window_size, nghttp2_session_get_remote_window_size, and returning NGHTTP2_ERR_WOULDBLOCK return code if it really wants.

**Comment by tatsuhiro-t** [comment]
I'm not sure how lighttpd is implemented, but it should provide aggregated WINDOW_UPDATE rather than sending WINDOW_UPDATE for each single DATA frame.  For instance, send WINDOW_UPDATE when window gets half of the max size.  Then it does not need to send 1 byte WINDOW_UPDATE.

**Comment by gstrauss** [comment]
> I'm not sure how lighttpd is implemented, but it should provide aggregated WINDOW_UPDATE rather than sending WINDOW_UPDATE for each single DATA frame.

I disagree.  lighttpd replenishes the sender's window as each DATA frame is received.  lighttpd does not know when the next TCP packet may arrive and there is non-zero latency between the client and server.  How long are you suggesting that lighttpd wait before sending a window update if no further DATA frames are forthcoming?

Instead, I think your suggest is a possible optimization: lighttpd could process all ready TCP packets and then send aggregate WINDOW_UPDATE before waiting to recv additional TCP packets.  However, that is an optimization, and lighttpd may receive only a single DATA frame before waiting for more TCP packets.  Therefore, your suggestion to provide aggregated WINDOW_UPDATE is not a solution to this issue.

> For instance, send WINDOW_UPDATE when window gets half of the max size. Then it does not need to send 1 byte WINDOW_UPDATE.

If a server is not using the HTTP/2 windows for flow control, the server can set very large values and avoid this issue.  However, if the HTTP/2 windows are being used for flow control, then your suggestion may slow things down due to delays replenishing the windows, and then latency sending the delayed WINDOW_UPDATE.  Would waiting for half the window size break an application?  That is specific to the application.  Are you suggesting doubling the window size to accommodate this?  That would reduce the utility of the windows for flow control.  Your suggestion to delay sending WINDOW_UPDATE until it reaches a watermark might be workable for some usage scenarios, but I do not feel that it is a generic workaround that the server should implement to avoid this client issue.  The client should know better if it is uploading a large file or sending a bunch of short messages, and can choose to wait for a larger window to be available or to send data without delay.

As I mentioned before, I think that the library libnghttp2 -- which may be managing the HTTP/2 window sizes for the application -- should consider implementing some defensive mechanisms or options to help the application avoid the degenerative behavior exhibited **by the application using libnghttp2**.

As an aside, lighttpd 1.4.65 HTTP/2 server implementation will change the default window size from 65535 to 65536, which is a simple and effective mitigation for specific instances of the issue reported here.

---

My instrumentation of curl shows me that libnghttp2 is calling the `read_callback` with a buffer of length 1 when either the session rwin or stream rwin is 1, so curl could use that instead of repeating calls to `nghttp2_session_get_stream_remote_window_size()` and `nghttp2_session_get_remote_window_size()`.

**Comment by tatsuhiro-t** [comment]
If server does not send 1 byte WINDOW_UPDATE, client would not send 1 byte DATA frame so often, so I suggested to aggregate them.  Basically WINDOW_SIZE should be size of BDP for optimal performance, if it is below that, the performance is lost anyway.  But if you do not like the idea, so be it.
Because nghttp2 provides all the means to application when to send data, it is up to the application how to behave in this kind of situation.  If this is curl issue, better to raise an issue on curl project.

**Comment by gstrauss** [comment]
> If server does not send 1 byte WINDOW_UPDATE, client would not send 1 byte DATA frame so often

We seem to have come to an end to the conversation since you are unwilling to discuss the **source** of the 1-byte DATA frames in the issue I have reported: the application which happens to use libnghttp2 to manage window sizes.

For my part -- and even though lighttpd **is not** the source of the issue -- I have modified lighttpd to adjust the server flow control window size, as I indicated above.

You seem unwilling to consider any changes to libnghttp2 to help the application avoid this degenerative behavior.  If that is the case, then please update the libnghttp2 documentation to highlight that libnghttp2 does generic window size management for minimal compliance with RFC 7540, and that the application is encouraged to fill the windows with all data available, or else the application should also participate in managing the window sizes to avoid degenerative behavior.  Thank you.

**Comment by gstrauss** [comment]
> Basically WINDOW_SIZE should be size of BDP for optimal performance, if it is below that, the performance is lost anyway.

NB: this issue is not about optimal performance.  This issue is about avoiding degenerative behavior ("pessimal performance").

If the goal was to increase performance without care for HTTP/2 flow control or memory usage, then the client and server could completely side-step **application level** HTTP/2 flow control by using maximal values for window sizes.  (off-topic: At much lower levels of the TCP stack, bandwidth delay product should be accommodated with larger kernel socket buffer sizes, either autotuned or explicit with `setsockopt()`.)

**Comment by gstrauss** [comment]
https://github.com/curl/curl/pull/8965 has been accepted into curl to address the degenerative behavior observed in curl.

There may be other applications which use libnghttp2 and which are vulnerable to devolving into this degenerative behavior.

@tatsuhiro-t if you are unwilling to consider any changes to libnghttp2 to help the application avoid this degenerative behavior (such as a configurable low-watermark under which the window is treated as empty), then please update the libnghttp2 documentation to highlight that libnghttp2 does generic window size management for minimal compliance with RFC 7540, and that the application is encouraged to fill the windows with all data available, or else the application should also participate in managing the window sizes to avoid degenerative behavior. Thank you.

**Comment by tatsuhiro-t** [comment]
If server does not send window_size_update = 1, this does not happen in the first place.
Best to avoid those servers. nghttp2 servers do not cause this issue.
When I suggest changing window update scheme of your server, you turned it down saying it is an optimization.  Then we just finished this matter.  Please close the issue.

**Comment by gstrauss** [comment]
Yes, servers do tend to be defensive and add workarounds for poorly written client software.  It would be nice if more client software was better written.

**Comment by gstrauss** [comment]
> nghttp2 servers do not cause this issue.

To be clear, I have provided documented evidence that a **client** using libnghttp2 is the source of this issue.  Since libnghttp2 handles the h2 window sizes for the client, libnghttp2 is implicated.  I have offered to provide a patch for libnghttp2 to help client applications using libnghttp2 avoid this degenerative behavior, but @tatsuhiro-t has declined.
